### PR TITLE
Test to ensure reader methods are observed

### PR DESCRIPTION
**Pull request by jarnaud**
## Motivation

We need to be sure that every time we add a method to the reader, the observability wrapper also implement the corresponding method.

## Solution

Since Go doesn't support inheritance, reflection cannot be used to detect if the wrapper is redefining the embedded methods (in both cases they appear as implemented by the embedding class).

This solutions analyze call stacks to check if the wrapper is used.

## Example

	// first 4 calls are the same for both wrapped and not wrapped, last call (4 or 7) is also the same.

	// NOT WRAPPED
	//caller 0 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/observability.TestOnRampObservedMethods.func1 file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go, line 27 (true)
	//caller 1 is github.com/stretchr/testify/mock.(*Mock).MethodCalled file /Users/jean/go/pkg/mod/github.com/stretchr/testify@v1.9.0/mock/mock.go, line 554 (true)
	//caller 2 is github.com/stretchr/testify/mock.(*Mock).Called file /Users/jean/go/pkg/mod/github.com/stretchr/testify@v1.9.0/mock/mock.go, line 466 (true)
	//caller 3 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/ccipdata/mocks.(*OnRampReader).IsSourceCursed file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/ccipdata/mocks/onramp_reader_mock.go, line 152 (true)
	//caller 4 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/observability.TestOnRampObservedMethods file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go, line 54 (true)

	// WRAPPED
	//caller 0 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/observability.TestOnRampObservedMethods.func1 file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go, line 27 (true)
	//caller 1 is github.com/stretchr/testify/mock.(*Mock).MethodCalled file /Users/jean/go/pkg/mod/github.com/stretchr/testify@v1.9.0/mock/mock.go, line 554 (true)
	//caller 2 is github.com/stretchr/testify/mock.(*Mock).Called file /Users/jean/go/pkg/mod/github.com/stretchr/testify@v1.9.0/mock/mock.go, line 466 (true)
	//caller 3 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/ccipdata/mocks.(*OnRampReader).IsSourceCursed file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/ccipdata/mocks/onramp_reader_mock.go, line 152 (true)
	//caller 4 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/observability.ObservedOnRampReader.IsSourceCursed.func1 file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/observability/onramp.go, line 49 (true)
	//caller 5 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/observability.withObservedInteraction[...] file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/observability/metrics.go, line 50 (true)
	//caller 6 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/observability.ObservedOnRampReader.IsSourceCursed file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/observability/onramp.go, line 48 (true)
	//caller 7 is github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/observability.TestOnRampObservedMethods file /Users/jean/workspaces/goland/ccip/core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go, line 61 (true)

### CODE REVIEW

File: core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	observed := ObservedOnRampReader{reader, metric}
+
+	// Test each method defined in the embedded type.
+	observedType := reflect.TypeOf(observed)
+	for i := 0; i < observedType.NumMethod(); i++ {
+
+		method := observedType.Method(i)
+		t.Run(fmt.Sprintf("observability_wrapper_%s", method.Name), func(t *testing.T) {
+
+			fmt.Printf("Testing method %s\n", method.Name)
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go]: can remove this line

File: core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+			mc := methodCalls[method.Name]
+			if mc.MethodName == "" {
+				assert.Fail(t, fmt.Sprintf("method %s not defined in methodCalls, please define it or exclude it.", method.Name))
+				return
+			}
+
+			// Defines mocked wrapped method behavior.
+			reader.On(mc.MethodName, mc.Arguments...).Maybe().Return(mc.Returns...).Run(func(args mock.Arguments) {
+				var observed = false
+				for i := 0; i < 8; i++ {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go]: what is `8`? can probably replace that with `len()`

File: core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+					pc, _, line, _ := runtime.Caller(i)
+					f := runtime.FuncForPC(pc)
+					//fmt.Printf("caller %d is %s file %v, line %d\n", i, f.Name(), file, line)
+					if strings.Contains(f.Name(), expectedWrapper) {
+						fmt.Printf("Method observed by wrapper at line %d\n", line)
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go]: can get rid of this line

File: core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+			// Retrieve method call from definition (fail if not present).
+			mc := methodCalls[method.Name]
+			if mc.MethodName == "" {
+				assert.Fail(t, fmt.Sprintf("method %s not defined in methodCalls, please define it or exclude it.", method.Name))
+				return
+			}
+
+			// Defines mocked wrapped method behavior.
+			reader.On(mc.MethodName, mc.Arguments...).Maybe().Return(mc.Returns...).Run(func(args mock.Arguments) {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/observability/onramp_observed_test.go]: maybe worths creating a function to simplify it a bit

**Review by dimkouv — APPROVED** [review]

**Review by mateusz-sekara — APPROVED** [review]

**Pull request merged**